File: app/models/member.rb
class Member < ApplicationRecord
  belongs_to :team
  belongs_to :user, optional: true, inverse_of: :memberships
  has_many :invitations, -> { pending }

  acts_as_url :name, url_attribute: :slug, limit: 128, scope: :team_id

  validates :name, presence: true, length: { maximum: 128 }
  validates :user_id, uniqueness: { scope: :team_id, allow_blank: true }
  validates :admin, absence: true, unless: :user?

  scope :admin, -> { where(admin: true) }

  alias to_param slug

  def user?
    user_id? || user.present?
  end
end

File: app/javascript/application/models/member.js
import React from 'react'
import Model from './model';

export default class Member extends Model {
  toString() {
    return this.name
  }

  url() {
    return `/teams/${this.team.slug}/people/${this.slug}`
  }

  set invitations(values) {
    this._invitations = (values || [])
  }

  get invitations() {
    return this._invitations || []
  }

  set invitation(value) {
    this._invitations = value ? [value] : []
  }

  get invitation() {
    return this.invitations[0]
  }
}
